neptune_optuna/impl/__init__.py

```diff
@@ -144,342 +175,484 @@ def __init__(self,
         self._log_plot_slice = log_plot_slice
         self._log_plot_intermediate_values = log_plot_intermediate_values
         self._log_plot_optimization_history = log_plot_optimization_history
+        self._target_names = target_names
+        self._namespaces = None
 
         run[INTEGRATION_VERSION_KEY] = __version__
 
     def __call__(self, study: optuna.Study, trial: optuna.trial.FrozenTrial):
-        self._log_trial(trial)
+        if self._namespaces is None:
+            self._namespaces = _get_namespaces(study, self._target_names)
+        self._log_trial(study, trial)
         self._log_trial_distributions(trial)
-        self._log_best_trials(study)
         self._log_study_details(study, trial)
         self._log_plots(study, trial)
         self._log_study(study, trial)
+        self._log_best_trials(study)
 
-    def _log_trial(self, trial):
-        _log_trials(self.run, [trial])
+    def _log_trial(self, study, trial):
+        _log_single_trial(self.run, study, trial=trial, namespaces=self._namespaces)
 
     def _log_trial_distributions(self, trial):
-        self.run['study/distributions'].log(trial.distributions)
+        self.run["study/distributions"].log(trial.distributions)
 
     def _log_best_trials(self, study):
-        self.run['best'] = _stringify_keys(_log_best_trials(study))
+        _log_best_trials(self.run, study, self._namespaces)
 
     def _log_study_details(self, study, trial):
         if trial._trial_id == 0:
```

Comment:
I know it's out of scope but in general using internal attributes like `_trial_id` is not a good idea. I found that there is `trial.number` that should work very similarly. https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.FrozenTrial.html#optuna.trial.FrozenTrial.number

Reply:
Ok

Reply:
So should I change it to `number` or not?